pkcs11/pkcs11_csr.c

```diff
@@ -0,0 +1,548 @@
+
+
+    /*------------------------------------------------------------------------*/
+    /* Loading the CA Certificate PEM File, used for the issuer in the new certificate */
+    /*------------------------------------------------------------------------*/
+    printf("Loading CA certificate\n");
+
+    file = fopen(caCertPemFile, "rb");
+    if (!file) {
+        printf("failed to open file: %s\n", caCertPemFile);
+        goto exit;
+    }
+    pemSz = fread(pemBuf, 1, LARGE_TEMP_SZ, file);
+    fclose(file);
+    printf("Read %d bytes from %s\n\n", pemSz, caCertPemFile);
+
+    ret = wc_CertPemToDer(pemBuf, pemSz, caCertBuf, LARGE_TEMP_SZ, CERT_TYPE);
+    if (ret == ASN_NO_PEM_HEADER) {
+        memcpy(caCertBuf, pemBuf, pemSz);
+        memset(pemBuf, 0, LARGE_TEMP_SZ);
+        caCertSz = pemSz;
+        printf("CA Cert file detected as DER\n\n");
+    }
+    else if (ret >= 0) {
+        caCertSz = ret;
+        ret = 0;
+        printf("Converted CA Cert PEM to DER %d bytes\n\n", caCertSz);
+    }
+    else {
+        goto exit;
+    }
+
+    /*------------------------------------------------------------------------*/
+    /* Load CSR PEM */
+    /*------------------------------------------------------------------------*/
+    printf("Loading CSR certificate\n");
+
+    file = fopen(csrPemFile, "rb");
+    if (!file) {
+        printf("failed to open file: %s\n", csrPemFile);
+        goto exit;
+    }
+    pemSz = fread(pemBuf, 1, LARGE_TEMP_SZ, file);
+    fclose(file);
+    printf("Successfully read %d bytes from %s\n\n", pemSz, csrPemFile);
+
+    ret = wc_CertPemToDer(pemBuf, pemSz, derBuf, LARGE_TEMP_SZ, CERTREQ_TYPE);
+    if (ret >= 0) {
+        derSz = ret;
+        ret = 0;
+    }
+    printf("Converted CSR Cert PEM to DER %d bytes\n", derSz);
+
+#ifdef HAVE_DECODEDCERT
+    /* Code for parsing a CSR to a DecodedCert struct */
+    /* Note: These are not public API's unless WOLFSSL_TEST_CERT or the compat 
+     * layer is enabled */
+    InitDecodedCert(&decoded, derBuf, derSz, HEAP_HINT);
+    ret = ParseCert(&decoded, CERTREQ_TYPE, NO_VERIFY, NULL);
+    if (ret != 0) goto exit;
+    printf("Loaded CSR to DecodedCert struct\n\n");
+    initDecode = 1;
+
+    /* Decode public key into csrKey */
+    printf("Decoding Public Key\n");
+    ret = wc_ecc_init(&csrKey);
+    if (ret != 0) goto exit;
+    initCsrKey = 1;
+
+    idx = 0;
+    ret = wc_EccPublicKeyDecode(decoded.publicKey, &idx, &csrKey,
+        decoded.pubKeySize);
+    if (ret != 0) goto exit;
+#endif
+
+    /*------------------------------------------------------------------------*/
+    /* Create a new certificate using SUBJECT information from CSR and ISSUER 
+     * from CA cert */
+    /*------------------------------------------------------------------------*/
+    wc_InitCert(&newCert);
+
+    newCert.isCA    = 0;
+    newCert.sigType = CTC_SHA256wECDSA;
+
+    printf("Setting certificate subject\n");
+#ifdef HAVE_DECODEDCERT
```

Comment:
Extract this code in to separate function and add comment to indicate customer specific changes required.